### PR TITLE
feat: add runtime css transformer

### PR DESCRIPTION
**Pull request by Saviio**
##### Checklist

- [x] `npm test` passes
- [x] tests are included
- [x] commit message follows commit guidelines
- [ ] documentation is changed or added

##### Description of change

qiankun offered `strictStyleIsolation` already, but it is implemented by shadow-dom which may casue other problems, we may need a more generic way to support `css isolation`. 

With this PR, a experimental feature is added, if you start qiankun with following args:
``` typescript
start({
  sandbox: {
    experimentalStyleIsolation: true
  }
})
```
the css style of sub app will be isolated by the runtime transformer, i.e. processor will add a specified attribute selector for every single css rule.
``` css
.app-main { ... }
↓ ↓ ↓
div[data-qiankun-react16] .app-main { ... }
↓ ↓ ↓
tag[data-qiankun-appname] sub-app-css-selector
```

#### These rules are not supported:
@keyframes, @font-face, @import, @page

#### Some component like Modal seems lost its style?
When option: experimentalStyleIsolation is true, all styles of sub app will be rewrite and qiankun will limit their scope of influence, therefore you have to manually add config code like this:
```typescript
// antd 
function render(props) {
  const { container } = props;
  const mountDOM = container ? container.querySelector('#root') : document.querySelector('#root')
  const rootDomGetter = () => mountDOM

  ReactDOM.render((
    <ConfigProvider getPopupContainer ={ rootDomGetter }>
      <App />
    </ConfigProvider>
  ), mountDOM);
}

``` 

#### My css is not be scoped?
currently we're not support tag: link, so if your css is be add by ```<link />```, this transformer will not process them.

### CODE REVIEW

File: src/sandbox/patchers/css.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+          const rules = arrayify<CSSRule>(sheet?.cssRules ?? []);
+          const css = this.rewrite(rules, prefix);
+
+          // eslint-disable-next-line no-param-reassign
+          styleNode.textContent = css;
+          // eslint-disable-next-line no-param-reassign
+          (styleNode as any)[ScopedCSS.ModifiedTag] = true;
+
+          // may cause leak?
+          mutator.disconnect();
```
Comment (Saviio) [inline src/sandbox/patchers/css.ts:70]: need cleanup to ensure wont cause dom leak?

File: src/sandbox/patchers/dynamicHeadAppend.ts [hunk 2]
```diff
@@ -64,6 +65,19 @@ export default function patch(
   singular = true,
 ): Freer {
   let dynamicStyleSheetElements: Array<HTMLLinkElement | HTMLStyleElement> = [];
+  let enableScopedCSS = frameworkConfiguration.experimental?.scopedCSS ?? false;
+
+  const mountedDOM = appWrapperGetter();
```
Comment (Saviio) [inline src/sandbox/patchers/dynamicHeadAppend.ts:70]: should handle undefined

File: src/loader.ts [hunk 3]
```diff
@@ -53,7 +54,25 @@ function createElement(appContent: string, strictStyleIsolation: boolean): HTMLE
     shadow.innerHTML = innerHTML;
   }
 
-  return appElement;
+  const cb: Function[] = [];
+  if (frameworkConfiguration.experimental?.scopedCSS ?? false) {
```
Comment (Saviio) [inline src/loader.ts:58]: use function arguments not global option object access?

File: src/loader.ts [hunk 4]
```diff
@@ -53,7 +54,25 @@ function createElement(appContent: string, strictStyleIsolation: boolean): HTMLE
...
+
+    for (let i = 0; i < len; i += 1) {
+      const styleNode = styleNodes[i];
+      if (styleNode.parentNode) {
+        styleNode.parentNode.removeChild(styleNode);
+      }
+
+      // trick, not elegant
+      // and may cause incorrect style order
+      cb.push(() => headNode?.appendChild(styleNode));
```
Comment (Saviio) [inline src/loader.ts:71]: need cr here

**Comment by simoncx** [comment]
问下这个模拟css隔离的功能什么时候可以merge？

**Comment by Saviio** [comment]
> 这种模拟隔离的方案，在多个应用同时加载时可行吗？

It should work if you run apps with unique name

**Comment by Saviio** [comment]
@kuitos PR is ready, have a look again?

File: src/sandbox/patchers/dynamicAppend.ts [hunk 5]
```diff
@@ -97,6 +99,10 @@ function getNewAppendChild(args: {
           if (!invokedByMicroApp) {
             return headOrBodyAppendChild.call(this, element) as T;
           }
+
```
Comment (kuitos) [inline src/sandbox/patchers/dynamicAppend.ts:102]:
```suggestion

```

File: src/sandbox/patchers/dynamicAppend.ts [hunk 6]
```diff
@@ -221,6 +227,10 @@ function getNewInsertBefore(args: {
           if (!invokedByMicroApp) {
             return headOrBodyInsertBefore.call(this, element, refChild) as T;
           }
+
```
Comment (kuitos) [inline src/sandbox/patchers/dynamicAppend.ts:230]:
```suggestion

```

File: src/sandbox/patchers/dynamicAppend.ts [hunk 7]
```diff
@@ -221,6 +226,9 @@ function getNewInsertBefore(args: {
           if (!invokedByMicroApp) {
             return headOrBodyInsertBefore.call(this, element, refChild) as T;
           }
+          if (isEnableScopedCSS(frameworkConfiguration)) {
```
Comment (kuitos) [inline src/sandbox/patchers/dynamicAppend.ts:229]:
```suggestion

          if (isEnableScopedCSS(frameworkConfiguration)) {
```

File: src/sandbox/patchers/dynamicAppend.ts [hunk 8]
```diff
@@ -97,6 +99,9 @@ function getNewAppendChild(args: {
           if (!invokedByMicroApp) {
             return headOrBodyAppendChild.call(this, element) as T;
           }
+          if (isEnableScopedCSS(frameworkConfiguration)) {
```
Comment (kuitos) [inline src/sandbox/patchers/dynamicAppend.ts:102]:
```suggestion

          if (isEnableScopedCSS(frameworkConfiguration)) {
```

File: src/utils.ts [hunk 9]
```diff
@@ -105,3 +106,15 @@ export function performanceMeasure(measureName: string, markName: string) {
...
+export function isEnableScopedCSS(opt: FrameworkConfiguration) {
+  if (typeof opt.sandbox !== 'object') {
+    return false;
+  }
+
+  if (opt.sandbox.strictStyleIsolation) {
+    return false;
+  }
+
+  return !!opt.sandbox.experimentalStyleIsolation;
```
[the remark this answers is not in the document]
Comment (Saviio) [reply src/utils.ts:119]: yes, but i think this way will be more readable?

File: src/sandbox/patchers/css.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    const css = this.rewrite(arrayify(rule.cssRules), prefix);
+    return `@supports ${rule.conditionText} {${css}}`;
+  }
+}
+
+const processor = new ScopedCSS();
+
+export const QiankunCSSRewriteAttr = 'data-qiankun';
+const process = (
+  appWrapper: () => HTMLElement,
```
[the remark this answers is not in the document]
Comment (Saviio) [reply src/sandbox/patchers/css.ts:140]: you mean just pass element as arg to this function?

File: src/utils.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kuitos) [reply src/utils.ts:119]: 👌

**Pull request merged**